pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -531,33 +564,31 @@ func (s *Server) createIstioRA(opts *caOptions) (ra.RegistrationAuthority, error
 		}
 
 		// File does not exist.
+		// TODO(costin): panic instead ?
+
 		if certSignerDomain == "" {
-			log.Infof("CA cert file %q not found, using %q.", caCertFile, defaultCACertPath)
-			caCertFile = defaultCACertPath
+			// We have no way to find the roots - this code path is not using MeshConfig.
+			log.Errorf("K8S RA %s without external-ca-cert secret mounted on /etc/external-ca-cert/root-cert.pem",
+				opts.ExternalCASigner)
+			return nil, err
+		}
+		log.Infof("K8S RA with domain %q using MeshConfig roots", certSignerDomain)
+	} else {
+		if certSignerDomain == "" {
+			log.Infof("K8S RA %s using roots from %s", opts.ExternalCASigner, caCertFile)
 		} else {
-			log.Infof("CA cert file %q not found - ignoring.", caCertFile)
-			caCertFile = ""
+			log.Infof("K8S RA with domain %s using roots from %s, MeshConfig roots ignored", certSignerDomain, caCertFile)
 		}
+		raOpts.CaCertFile = caCertFile
 	}
 
-	if s.kubeClient == nil {
-		return nil, fmt.Errorf("kubeClient is nil")
-	}
-	raOpts := &ra.IstioRAOptions{
-		ExternalCAType:   opts.ExternalCAType,
-		DefaultCertTTL:   workloadCertTTL.Get(),
-		MaxCertTTL:       maxWorkloadCertTTL.Get(),
-		CaSigner:         opts.ExternalCASigner,
-		CaCertFile:       caCertFile,
-		VerifyAppendCA:   true,
-		K8sClient:        s.kubeClient.Kube(),
-		TrustDomain:      opts.TrustDomain,
-		CertSignerDomain: opts.CertSignerDomain,
-	}
-	raServer, err := ra.NewIstioRA(raOpts)
+	raServer, err := ra.NewKubernetesRA(raOpts)
```

Comment:
Why are we calling this the kubernetesRA instead of the istioRA?

Reply:
There is nothing Istio specific to it - we are using K8S standard API to sign certificates. The signers are not aware or care about Istio, and Istio doesn't define the protocol. 

IstioRA would be if we were using the Istio GRPC signer method to chain - but that is not a good idea for many reasons. 

Technically the K8S signing is not a good fit for RA either - it is closer to ACME or classical signing, where the CSR is published by the original, intended user and the 'approver' and 'signer' have full information and can apply their own policies. As we are using it - the signer has no way to apply policies, since it sees Istiod as the source of the CSR.